compiler/rustc_codegen_ssa/src/base.rs

```diff
@@ -156,7 +156,8 @@ pub fn unsized_info<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
         {
             let old_info =
                 old_info.expect("unsized_info: missing old info for trait upcasting coercion");
-            if data_a.principal_def_id() == data_b.principal_def_id() {
+            let b_principal_def_id = data_b.principal_def_id();
+            if data_a.principal_def_id() == b_principal_def_id || b_principal_def_id.is_none() {
```

Comment:
What would happen if you did not do the `is_none` check here? Does `supertrait_vtable_slot` return a bogus vtable entry or ICE or something like that?

Reply:
Below, if `supertrait_vtable_slot` returns `None` we also just keep the old pointer. So is this here just an optimization to avoid calling `supertrait_vtable_slot`?

Reply:
> is this here just an optimization to avoid calling supertrait_vtable_slot?

@compiler-errors wrote this code, but I believe that is correct

Reply:
FWIW I was asking this while writing https://github.com/rust-lang/rust/pull/127856, and that PR now has a bunch of assertions double-checking my understanding of the relevant invariants here.